### PR TITLE
Support SOF on imx8ulp

### PR DESCRIPTION
**Pull request by TangleZ**

### CODE REVIEW

File: src/include/sof/lib/dai.h [hunk 1]
```diff
@@ -417,6 +420,18 @@ static inline int dai_get_fifo(struct dai *dai, int direction,
...
+/**
+ * \brief Get Digital Audio dma srcid
+ */
+static inline int dai_get_srcid(struct dai *dai, int direction,
+			       int stream_id)
+{
+	int ret = dai->drv->ops.get_srcid(dai, direction, stream_id);
+
+
+	return ret;
```
Comment (marc-hb) [inline src/include/sof/lib/dai.h:432]: checkpatch warning?

File: src/include/sof/platform.h [hunk 2]
```diff
@@ -47,6 +47,28 @@ int platform_init(struct sof *sof);
...
+		break;
+	case LOCAL_TO_HOST:
+		*addr = MEM_RESERVED + (*addr - SDRAM0_BASE);
+		break;
+	};
+}
+#else
+static inline void convert_addr(int type, uint32_t *addr)
+{
+	return;
```
Comment (marc-hb) [inline src/include/sof/platform.h:68]: Same comment as above.

File: src/platform/imx8ulp/platform.c [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	.version = {
+		.hdr.size = sizeof(struct sof_ipc_fw_version),
+		.micro = SOF_MICRO,
+		.minor = SOF_MINOR,
+		.major = SOF_MAJOR,
+#ifdef DEBUG_BUILD
+		/* only added in debug for reproducability in releases */
+		.build = SOF_BUILD,
+		.date = __DATE__,
+		.time = __TIME__,
```
Comment (marc-hb) [inline src/platform/imx8ulp/platform.c:56]: This was changed in 9f8cce15224, can you align?

The joys of copy/paste/diverge...

File: src/drivers/imx/interrupt-noirq.c [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+DECLARE_TR_CTX(noirq_i_tr, SOF_UUID(noirq_imx_uuid), LOG_LEVEL_INFO);
+
+int irqstr_get_sof_int(int irqstr_int)
+{
+	return irqstr_int;
+}
+
+void platform_interrupt_init(void)
+{
+	return;
```
Comment (marc-hb) [inline src/drivers/imx/interrupt-noirq.c:34]: I find an empty blocks clearer and faster to read. I think they're more common here.

File: src/drivers/imx/interrupt-noirq.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+void platform_interrupt_set(uint32_t irq)
+{
+	arch_interrupt_set(irq);
+	return;
```
Comment (marc-hb) [inline src/drivers/imx/interrupt-noirq.c:40]: Similarly, returning nothing seems to just add some noise. I've never seen this in this codebase.

**Review by marc-hb — COMMENTED** [review]
On one hand you're making the new IMX_NOIRQ and IMX_IRQ visible, but on the other hand you enforce them with `select`. Do you want users to be able to decide IMX_IRQ freely? If not then make them not visible.

I'm also confused because there seems to be only two possibilities: either `interrupt-irqsteer.c` or `interrupt-noirq.c`. Two possibilities should mean only one boolean but this PR is raising the total to 3 booleans: IMX_NOIRQ, IMX_IRQ and IMX_INTERRUPT_IRQSTEER. That seems too complicated.

Will there be more than two possibilities in the future? In that case maybe it shouldn't be a boolean.

https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
```
select should be used with care. select will force
	a symbol to a value without visiting the dependencies.
	By abusing select you are able to select a symbol FOO even
	if FOO depends on BAR that is not set.
In general **use select only for non-visible symbols**
	(no prompts anywhere) and for symbols with no dependencies
```
Also: https://docs.zephyrproject.org/1.14.0/guides/kconfig/index.html#select-recommendations

File: src/drivers/imx/edma.c [hunk 6]
```diff
@@ -25,6 +25,21 @@ DECLARE_SOF_UUID("edma", edma_uuid, 0x3d73a110, 0x0930, 0x457f,
...
+{
+	if (channel->direction == DMA_DIR_MEM_TO_DEV)
+		convert_addr(LOCAL_TO_HOST, sbase);
+	else
+		convert_addr(LOCAL_TO_HOST, dbase);
+}
+#else
+static void chan_addr_convert(struct dma_chan_data *channel, uint32_t *sbase, uint32_t *dbase)
+{
+	return;
```
[the remark this answers is not in the document]
Comment (TangleZ) [reply src/drivers/imx/edma.c:39]: Hi marc-hb, Thanks for your review my patches! I will update this.

File: src/platform/imx8ulp/include/platform/drivers/interrupt.h [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+
+/* IRQ numbers */
+#define IRQ_DSP_SYS_NMI      0
+#define IRQ_NUM_SOFTWARE0    1
+#define IRQ_NUM_TIMER0       2
+#define IRQ_NUM_TIMER1       3
+#define DSP_Profiling_IRQ_L3 4
+#define DSP_Wer_L3 5
+#define DMA2_c0_ErrTranc 6
```
Comment (paulstelian97) [inline src/platform/imx8ulp/include/platform/drivers/interrupt.h:23]: This seems out of place with the regular naming convention. If you aren't using or handling them from SOF, you don't need to define them at all anyway.

File: src/platform/imx8ulp/include/platform/drivers/interrupt.h [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+#define IRQSTR_CH_SET(n)		(0x44 + 0x04 * (15 - (n)))
+#define IRQSTR_CH_STATUS(n)		(0x84 + 0x04 * (15 - (n)))
+#define IRQSTR_MASTER_DISABLE		0xC4
+#define IRQSTR_MASTER_STATUS		0xC8
+
+#define IRQSTR_RESERVED_IRQS_NUM	32
+#define IRQSTR_IRQS_NUM			512
+#define IRQSTR_IRQS_REGISTERS_NUM	16
+#define IRQSTR_IRQS_PER_LINE		64
+*/
```
Comment (paulstelian97) [inline src/platform/imx8ulp/include/platform/drivers/interrupt.h:63]: Commented out code that has no reason to stay. Just delete it (we will always have it for reference in the other platforms anyway).

File: src/platform/imx8ulp/include/platform/drivers/interrupt.h [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+/* no irqstr in 8ulp */
+#define IRQSTR_CH_MASK(n)		(0x04 + 0x04 * (15 - (n)))
```
Comment (paulstelian97) [inline src/platform/imx8ulp/include/platform/drivers/interrupt.h:66]: Are these macros even necessary in the first place? If so, at least make this one trivial (say, define it as "n")

File: src/platform/imx8ulp/include/platform/lib/clk.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <stdint.h>
+
+struct sof;
+
+#define CLK_CPU(x)	(x)
+
+#define CPU_DEFAULT_IDX		0
+
+#ifdef CONFIG_IMX8
```
Comment (paulstelian97) [inline src/platform/imx8ulp/include/platform/lib/clk.h:21]: ULP should match neither of these, so remove the dead code.

File: src/platform/imx8ulp/include/platform/lib/cpu.h [hunk 11]
```diff
@@ -0,0 +1,27 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2019 Intel Corporation. All rights reserved.
+ *
+ * Author: Tomasz Lauda <tomasz.lauda@linux.intel.com>
```
Comment (paulstelian97) [inline src/platform/imx8ulp/include/platform/lib/cpu.h:5]: I'd believe this file you should update the copyright and author yourself?

File: src/platform/imx8ulp/include/platform/lib/dai.h [hunk 12]
```diff
@@ -0,0 +1,26 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright 2019 NXP
+ *
+ * Author: Daniel Baluta <daniel.baluta@nxp.com>
```
Comment (paulstelian97) [inline src/platform/imx8ulp/include/platform/lib/dai.h:5]: Ditto, the bulk of the contents is yours. Update copyright year.

File: src/platform/imx8ulp/include/platform/lib/dma.h [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+#ifndef __PLATFORM_LIB_DMA_H__
+#define __PLATFORM_LIB_DMA_H__
+
+#define PLATFORM_NUM_DMACS	3
+
+/* max number of supported DMA channels */
+#define PLATFORM_MAX_DMA_CHAN	32
+
+#define DMA_ID_EDMA2	0
+#define DMA_ID_HOST	1
```
Comment (paulstelian97) [inline src/platform/imx8ulp/include/platform/lib/dma.h:19]: PLATFORM_NUM_DMACS is 3 but we have only two IDs, this doesn't look right. Also ULP has EDMA?

File: src/platform/imx8ulp/lib/clk.c [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+#include <sof/common.h>
+#include <sof/lib/clk.h>
+#include <sof/lib/cpu.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/notifier.h>
+#include <sof/sof.h>
+#include <sof/spinlock.h>
+
+const struct freq_table platform_cpu_freq[] = {
+#ifdef CONFIG_IMX8
```
Comment (paulstelian97) [inline src/platform/imx8ulp/lib/clk.c:17]: Ditto: Neither should match on ULP, add your own entries.

File: src/include/sof/platform.h [hunk 15]
```diff
@@ -47,6 +47,28 @@ int platform_init(struct sof *sof);
 
 /** @}*/
 
+#define HOST_TO_LOCAL 0
+#define LOCAL_TO_HOST 1
+
+#ifdef CONFIG_IMX8ULP
+static inline void convert_addr(int type, uint32_t *addr)
+{
```
Comment (dbaluta) [inline src/include/sof/platform.h:55]: This function doesn't belong to this commit. Current commit is about Messaging Unit.

File: src/platform/imx8ulp/lib/clk.c [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+	{ 666000000, 666000 },
+#else /* CONFIG_IMX8X */
+	{ 640000000, 640000 },
```
Comment (paulstelian97) [inline src/platform/imx8ulp/lib/clk.c:20]: Do the IMX8X entries work on ULP? If not, add your own.

File: src/platform/imx8ulp/lib/dai.c [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+	.plat_data = {
+		.base = SAI_5_BASE,
+		.fifo[SOF_IPC_STREAM_PLAYBACK] = {
+			.offset		= SAI_5_BASE + REG_SAI_TDR0,
+			/* use depth to model the HW FIFO size:
+			 * each channel includes a 64 x 32 bit FIFO
+			 * that can be accessed using Transmit or
+			 * Receive Data Registers
+			 */
+			.depth		= 16,  /* in 4 bytes words */
```
Comment (paulstelian97) [inline src/platform/imx8ulp/lib/dai.c:29]: Mismatch between 16 and 64 in the comment?

File: src/audio/dai.c [hunk 18]
```diff
@@ -393,7 +393,9 @@ static void dai_data_config(struct comp_dev *dev)
 		dd->stream_id = dd->dai_config->alh.stream_id;
```
Comment (dbaluta) [inline src/audio/dai.c:393]: `src` prefix is redundant in the commit message. Please use `git log src/audio/dai.c` and see what other have used for the prefixes.

Also, the commit subject doesn't make much sense. Please make the commit message to be an action that this commit is doing on the existing code.

e.g dai: imx: Add support for i.MX8ULP

Also, please read again https://chris.beams.io/posts/git-commit/#seven-rules

File: src/audio/dai.c [hunk 19]
```diff
@@ -393,7 +393,9 @@ static void dai_data_config(struct comp_dev *dev)
...
 		break;
 	case SOF_DAI_IMX_SAI:
-		COMPILER_FALLTHROUGH;
+		dd->config.burst_elems =
+			dd->dai->plat_data.fifo[dai->direction].depth;
+		//COMPILER_FALLTHROUGH;
```
Comment (paulstelian97) [inline src/audio/dai.c:398]: This change doesn't look good at all. You're falling through and running the same statement twice. What's the point in the first place?

File: src/include/sof/drivers/sai.h [hunk 20]
```diff
@@ -247,15 +249,28 @@
  * configured for an internal bit clock.
  * The division value is (DIV + 1) * 2.
  */
+#ifndef CONFIG_IMX8ULP
 #define SAI_CLOCK_DIV		0x7
+#else
+#define SAI_CLOCK_DIV		0x5
```
Comment (paulstelian97) [inline src/include/sof/drivers/sai.h:255]: CLOCK_DIV should be 5 on NON-ULP platforms where it was 7 before? I feel like the logic is flipped...

File: src/drivers/imx/edma.c [hunk 21]
```diff
@@ -25,6 +25,21 @@ DECLARE_SOF_UUID("edma", edma_uuid, 0x3d73a110, 0x0930, 0x457f,
 
 DECLARE_TR_CTX(edma_tr, SOF_UUID(edma_uuid), LOG_LEVEL_INFO);
```
Comment (dbaluta) [inline src/drivers/imx/edma.c:26]: Please always explain why the commit is needed. 'Prepare for 8ulp' is not really self-explanatory. Also the code is not self explanatory.

When creating the commit please explain everything as you would want other to explain their code so that you can understand it.

File: src/include/sof/drivers/edma.h [hunk 22]
```diff
@@ -95,4 +96,18 @@
 #define EDMA0_SAI_CHAN_RX_IRQ	349
 #define EDMA0_SAI_CHAN_TX_IRQ	349
 
+/* 8ulp chans */
+#define EDMA2_CHAN0    0
```
Comment (paulstelian97) [inline src/include/sof/drivers/edma.h:100]: 8ULP specific defines should be in an 8ULP-only header. You can make your own platform edma.h header to have these defines.

File: src/drivers/Kconfig [hunk 23]
```diff
@@ -20,6 +20,20 @@ config DUMMY_DMA
...
+	  Select this to enable support for i.MX interrupt with irq.
+
+config IMX_NOIRQ
+	bool "i.MX NO IRQ"
+	default n
+	depends on IMX
+	help
+	  Select this to enable support for i.MX interrupt without irq.
+
 config IPC_POLLING
```
Comment (dbaluta) [inline src/drivers/Kconfig:37]: I think for now IMX platforms fall into 2 categories
- with irqsteer -> people should select CONFIG_IMX_INTERRUPT_IRQSTEER
- withouht irqsteer -> I think here we should add an option named CONFIG_IMX_GENERIC_INTERRUPT

and that's all. IMX_IRQ and IMX_NOIRQ are not really two options. Is just one option and the other one is negation of first.

File: tools/topology/sof-imx8ulp-btsco.m4 [hunk 24]
```diff
@@ -0,0 +1,82 @@
+#
+# Topology for i.MX8ULP board with wm8960 codec
```
Comment (dbaluta) [inline tools/topology/sof-imx8ulp-btsco.m4:2]: Please explain here what are the DAIs, what is the codec. Some information about the audio topology written in words is more than necessary. 

I bet no one knows what does btsco means.

File: scripts/qemu-check.sh [hunk 25]
```diff
@@ -139,7 +139,7 @@ do
 			SHM_MBOX=qemu-bridge-hp-sram-mem
 			has_rom=true
 			;;
-		imx8 | imx8x | imx8m)
+		imx8 | imx8x | imx8m | imx8ulp)
 			READY_IPC="00 00 00 00 00 00 04 c0"
```
Comment (dbaluta) [inline scripts/qemu-check.sh:143]: this is for qemu support. I don't think we have qemu support for hifi4 integration in imx8ulp. So, I think we can skip it for now

File: src/include/sof/drivers/edma.h (hunk 22 above)
Comment (dbaluta) [reply src/include/sof/drivers/edma.h:100]: @paulstelian97 @TangleZ I don't think we want another header just for this. Maybe we can add 8ULP in the name of changes.

E.g EDMA2_8ULP_CHAN0 0

File: src/drivers/Kconfig (hunk 23 above)
Comment (marc-hb) [reply src/drivers/Kconfig:37]: Great, so it can be made a single boolean then.
